analyzers/src/SonarAnalyzer.Common/Trackers/ArgumentTracker.cs

```diff
@@ -0,0 +1,86 @@
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+* Lesser General Public License for more details.
+*
+* You should have received a copy of the GNU Lesser General Public License
+* along with this program; if not, write to the Free Software Foundation,
+* Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+*/
+
+namespace SonarAnalyzer.Helpers.Trackers;
+
+public abstract class ArgumentTracker<TSyntaxKind> : SyntaxTrackerBase<TSyntaxKind, ArgumentContext>
+    where TSyntaxKind : struct
+{
+    protected abstract RefKind? ArgumentRefKind(SyntaxNode argumentNode);
+    protected abstract IReadOnlyCollection<SyntaxNode> ArgumentList(SyntaxNode argumentNode);
+    protected abstract int? Position(SyntaxNode argumentNode);
+    protected abstract bool InvocationFitsMemberKind(SyntaxNode invokedExpression, MemberKind memberKind);
+    protected abstract bool InvokedMemberFits(SemanticModel model, SyntaxNode invokedExpression, MemberKind memberKind, Func<string, bool> invokedMemberNameConstraint);
+
+    protected override ArgumentContext CreateContext(SonarSyntaxNodeReportingContext context) =>
+        new(context);
+
+    public Condition MatchArgument(ArgumentDescriptor descriptor) =>
+        context =>
+        {
+            if (context.Node is { } argumentNode
+                && argumentNode is { Parent.Parent: { } invoked }
+                && SyntacticChecks(context.SemanticModel, descriptor, argumentNode, invoked)
+                && (descriptor.InvokedMemberNodeConstraint?.Invoke(context.SemanticModel, Language, invoked) ?? true)
+                && MethodSymbol(context.SemanticModel, invoked) is { } methodSymbol
+                && Language.MethodParameterLookup(invoked, methodSymbol).TryGetSymbol(argumentNode, out var parameter)
+                && ParameterFits(parameter, descriptor.ParameterConstraint, descriptor.InvokedMemberConstraint))
+            {
+                context.Parameter = parameter;
+                return true;
+            }
+            return false;
+        };
+
+    private IMethodSymbol MethodSymbol(SemanticModel model, SyntaxNode invoked) =>
+        model.GetSymbolInfo(invoked).Symbol switch
+        {
+            IMethodSymbol x => x,
+            IPropertySymbol propertySymbol => Language.Syntax.IsWrittenTo(invoked, model, CancellationToken.None)
+                ? propertySymbol.SetMethod
+                : propertySymbol.GetMethod,
+            _ => null,
+        };
+
+    private bool SyntacticChecks(SemanticModel model, ArgumentDescriptor descriptor, SyntaxNode argumentNode, SyntaxNode invokedExpression) =>
+        InvocationFitsMemberKind(invokedExpression, descriptor.MemberKind)
+        && (descriptor.RefKind is not { } expectedRefKind || ArgumentRefKind(argumentNode) is not { } actualRefKind || actualRefKind == expectedRefKind)
```

Comment:
Maybe I'm missing something but I read this as:
- [1] if there is no constraint on the descriptor -> true
- [2] if there is no ref kind defined on the node -> true
- [3] if there is one defined on the node, it should be equal to the one defined on the descriptor.

Regarding point [2], what if there is a RefKind defined on the descriptor, shouldn't it return false?

Reply:
Regarding 2) RefKind of the node is either:
* RefKind.Out -> `out` is present
* RefKind.Ref -> `ref` is present
* RefKind.None -> no RefKind is specified
* null -> The argument node does not support RefKind

Reply:
Discussed offline, this handles a particular VB case, where the ref kind does not have to be specified on the call side.

```csharp
|| ArgumentRefKind(argumentNode) is not { } actualRefKind // This is to support VB which does not require the ref/out keywords on the call side.
```